transports/quic/src/transport.rs

```diff
@@ -0,0 +1,297 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the [`Transport`] trait for QUIC.
+//!
+//! Combines all the objects in the other modules to implement the trait.
+
+use crate::{endpoint::Endpoint, in_addr::InAddr, muxer::QuicMuxer, upgrade::Upgrade};
+
+use futures::prelude::*;
+use futures::stream::StreamExt;
+
+use if_watch::IfEvent;
+
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerEvent, TransportError},
+    PeerId, Transport,
+};
+use std::task::{Context, Poll};
+use std::{net::SocketAddr, pin::Pin, sync::Arc};
+
+// We reexport the errors that are exposed in the API.
+// All of these types use one another.
+pub use crate::connection::Error as Libp2pQuicConnectionError;
+pub use quinn_proto::{
+    ApplicationClose, ConfigError, ConnectError, ConnectionClose, ConnectionError,
+    TransportError as QuinnTransportError, TransportErrorCode,
+};
+
+/// Wraps around an `Arc<Endpoint>` and implements the [`Transport`] trait.
+///
+/// > **Note**: This type is necessary because Rust unfortunately forbids implementing the
+/// >           `Transport` trait directly on `Arc<Endpoint>`.
+#[derive(Debug, Clone)]
+pub struct QuicTransport {
+    endpoint: Arc<Endpoint>,
+    /// The IP addresses of network interfaces on which the listening socket
+    /// is accepting connections.
+    ///
+    /// If the listen socket listens on all interfaces, these may change over
+    /// time as interfaces become available or unavailable.
+    in_addr: InAddr,
+}
+
+impl QuicTransport {
+    pub fn new(endpoint: Arc<Endpoint>) -> Self {
+        let in_addr = InAddr::new(endpoint.local_addr.ip());
+        Self { endpoint, in_addr }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Error while trying to reach a remote.
+    #[error("{0}")]
+    Reach(ConnectError),
+    /// Error after the remote has been reached.
+    #[error("{0}")]
+    Established(Libp2pQuicConnectionError),
+    /// Error while working with IfWatcher.
+    #[error("{0}")]
+    IfWatcher(std::io::Error),
+}
+
+impl Transport for QuicTransport {
+    type Output = (PeerId, QuicMuxer);
+    type Error = Error;
+    // type Listener = Pin<
+    //     Box<dyn Stream<Item = Result<ListenerEvent<Upgrade, Self::Error>, Self::Error>> + Send>,
+    // >;
+    type Listener = Self;
```

Comment:
With #2652 this won't be needed anymore. So I think `listen_on` can just be a noop. Eventually we might also want to allow multiple listeners / endpoints, but I think this is out of scope for this PR.

Reply:
I did a quick draft of how the implementation can be adapted to the transport changes: b239de6. It does not change much, but at least improves the listening logic a bit.
I thought about the listening logic a bit more and contrary to what I wrote above, I do think we already have to support multiple listeners / endpoints in the first version. 
Once #2652 is merged I am planning to work on a PR to this implementation to adapt it both, the listener changes and the StreamMuxer refactor (in particular #2707). Just to avoid duplicating work: are you currently working on this @kpp?

Reply:
I am working on https://github.com/libp2p/rust-libp2p/pull/2707 only.